core/optaplanner-constraint-streams/src/main/java/org/optaplanner/constraint/streams/bavet/bi/JoinBiNode.java

```diff
@@ -76,12 +75,9 @@ public void insertA(UniTuple<A> tupleA) {
         Object[] indexProperties = mappingA.apply(tupleA.factA);
         tupleA.store[inputStoreIndexA] = indexProperties;
 
-        Map<UniTuple<B>, Set<BiTuple<A, B>>> tupleABSetMapB = indexerB.get(indexProperties);
-        // Use standard initial capacity (16) to grow into, unless we already know more is probably needed
-        Set<BiTuple<A, B>> tupleABSetA = new LinkedHashSet<>(Math.max(16, tupleABSetMapB.size()));
+        Set<BiTuple<A, B>> tupleABSetA = new LinkedHashSet<>();
         indexerA.put(indexProperties, tupleA, tupleABSetA);
-
-        tupleABSetMapB.forEach((tupleB, tupleABSetB) -> {
+        indexerB.visit(indexProperties, (tupleB, tupleABSetB) -> {
```

Comment:
Before, `indexer.get()` would iterate over everything, and create the `tupleABSetMapB` map. This map would then be iterated over once again, and thrown away. This is a waste of both time and memory, especially as the indexes grow.

Now, instead of `indexer.get()`, we use `indexer.visit()` - no map is ever created, and the iteration is only performed once.

We lose the ability to size `tupleABSetA`, as we never have the final map, but that doesn't matter. Even without proper sizing, the smallest score calc speed increase my non-scientific quick benchmark of nqueens got was 10 %.